core/server/worker/src/main/java/alluxio/worker/block/UfsInputStreamManager.java

```diff
@@ -0,0 +1,300 @@
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.worker.block;
+
+import alluxio.Configuration;
+import alluxio.PropertyKey;
+import alluxio.underfs.SeekableUnderFileInputStream;
+import alluxio.underfs.UnderFileSystem;
+import alluxio.underfs.options.OpenOptions;
+import alluxio.util.IdUtils;
+
+import com.google.common.base.Preconditions;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.RemovalListener;
+import com.google.common.cache.RemovalListeners;
+import com.google.common.cache.RemovalNotification;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * An under filesystem input stream manager that can cache seekable input streams for future reuse.
+ * The manager caches the input streams, with a time-based eviction policy of expiration configured
+ * in {@link PropertyKey.WORKER_UFS_INSTREAM_CACHE_EXPIRE_MS}.
+ *
+ * An under storage file maybe opened with multiple input streams at the same time. The manager uses
+ * {@link UfsInputStreamIds} to track the in-use input stream and the available ones. The manager
+ * closes the input streams after they are expired and not in-use anymore.
+ */
+@ThreadSafe
+public class UfsInputStreamManager {
+  private static final Logger LOG = LoggerFactory.getLogger(UfsInputStreamManager.class);
+
+  /**
+   * A map from the ufs file name to the metadata of the input streams. Synchronization is needed to
+   * access this map.
+   */
+  private final HashMap<String, UfsInputStreamIds> mFileToInputStreamIds;
+  /**
+   * Cache of the input streams, from the input stream id to the input stream.
+   */
+  @GuardedBy("mFileToInputStreamIds")
+  private final Cache<Long, SeekableUnderFileInputStream> mUnderFileInputStreamCache;
+  /**
+   * Thread pool for asynchronously removing the expired input streams.
+   */
+  private final ExecutorService mRemovalThreadPool;
+
+  /**
+   * Creates a new {@link UfsInputStreamManager}.
+   */
+  public UfsInputStreamManager() {
+    mFileToInputStreamIds = new HashMap<>();
+    mRemovalThreadPool = Executors.newFixedThreadPool(2);
```

Comment:
What factors influence the optimal pool size? Should the threads be daemon threads?

Reply:
This is a simple pool for cache maintenance work and the removal work is light,  so I use small number of threads. Why are daemon threads are needed? I think normal threads should be sufficient.

Reply:
Normal threads will prevent the JVM from exiting